x/cash/dynamicfee_test.go

```diff
@@ -0,0 +1,244 @@
+		t.Run(testName, func(t *testing.T) {
+			auth := helpers.Authenticate(tc.signers...)
+			bucket := NewBucket()
+			ctrl := NewController(bucket)
+			h := NewDynamicFeeDecorator(auth, ctrl)
+
+			tx := &txMock{info: &FeeInfo{Fees: &tc.txFee}}
+
+			db := store.MemStore()
+
+			gconf.SetValue(db, GconfCollectorAddress, collectorAddr)
+			gconf.SetValue(db, GconfMinimalFee, tc.minimumFee)
+
+			ensureWallets(t, db, tc.initWallets)
+
+			cache := db.CacheWrap()
+
+			cRes, err := h.Check(nil, cache, tx, tc.handler)
+			if !errors.Is(tc.wantCheckErr, err) {
+				t.Fatalf("got check error: %v", err)
+			}
+			if tc.wantGasPayment != cRes.GasPayment {
+				t.Errorf("gas payment: %d", cRes.GasPayment)
+			}
+
+			assertCharged(t, cache, ctrl, tc.wantCheckTxFee)
+
+			ensureWallets(t, cache, tc.updateWallets)
+
+			// If the check failed, deliver must not be called.
+			if tc.wantCheckErr != nil {
+				return
+			}
+
+			cache.Discard()
+
+			if _, err = h.Deliver(nil, cache, tx, tc.handler); !errors.Is(tc.wantDeliverErr, err) {
+				t.Fatalf("got deliver error: %v", err)
+			}
+
+			assertCharged(t, cache, ctrl, tc.wantDeliverTxFee)
+		})
+	}
+}
+
+var helpers x.TestHelpers
+
+// ensureWallets persist state of given wallet objects in the database. If
+// a wallet already exist it is overwritten.
+func ensureWallets(t *testing.T, db weave.KVStore, wallets []orm.Object) {
+	t.Helper()
+
+	bucket := NewBucket()
+	for i, w := range wallets {
+		if err := bucket.Save(db, w); err != nil {
+			t.Fatalf("cannot set %d wallet: %s", i, err)
+		}
+	}
+}
+
+// assertCharged check that given account was charged according to the fee
+// configuration.
+func assertCharged(t *testing.T, db weave.KVStore, ctrl Controller, want x.Coin) {
+	t.Helper()
+
+	minimumFee := gconf.Coin(db, GconfMinimalFee)
+	collectorAddr := gconf.Address(db, GconfCollectorAddress)
+
+	switch chargedFee, err := ctrl.Balance(db, collectorAddr); {
+	case err == nil:
+		wantTx := x.Coins{&want}
+		if !wantTx.Equals(chargedFee) {
+			t.Errorf("charged fee: %v", chargedFee)
+		}
+	case errors.Is(errors.ErrNotFound, err):
+		if minimumFee.IsZero() {
+			// Minimal fee is zero so the collector account is zero
+			// as well (not even created). All good.
+		} else {
+			if want.IsZero() {
+				// This is a weird case when a transaction was
+				// submitted but the signer does not have
+				// enough funds to pay the minimum (anty spam)
+				// fee.
+			} else {
+				t.Error("no fee charged")
+			}
+		}
+	default:
+		t.Errorf("cannot check collector account balance: %s", err)
+	}
+}
+
+type txMock struct {
+	weave.Tx
+	FeeTx
+	info *FeeInfo
+}
+
+func (m *txMock) GetFees() *FeeInfo {
+	return m.info
+}
+
+// Declare a unique error that can be matched in tests. This error is declared
+// only in tests so there is no way it can be returned by the implementation by
+// an accident.
+var ErrTestingError = errors.Register(123456789, "testing error")
+
+type handlerMock struct {
+	weave.Handler
```

Comment:
Why embed this and then never set it above?
Seems like a trick to get this to compile and fulfill the interface, but at the risk of a nil interface error down the line. Please remove and use

`var _ weave.Handler = (*mockHandler)(nil)` to ensure it fits the interface

Reply:
Those two constructs serve a different purpose.

What I did, as you wrote, is for the compiler to compile. The compiler will think that `handleMock` is implementing `weave.Handler` interface.
Code will panic if a method of a `weave.Handler` is used and not explicitly implemented. This is acceptable for tests.

The later test of the interface is a different thing. This is a way to enforce during the compilation time that the structure implements the interface. The compilation will fail if it does not. This construct is for testing not extending.

Does this make sense?

Reply:
Yes, but since you never actually use the weave.Handler to extend, it seemed pointless, and I removed it.